dns_sd.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * libiio - Library for interfacing industrial I/O (IIO) devices
+ *
+ * Copyright (C) 2014-2020 Analog Devices, Inc.
+ * Author: Robin Getz <robin.getz@analog.com>
+ *         Matej Kenda <matejken@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * Some of this is insipred from libavahi's example:
+ * https://avahi.org/doxygen/html/client-browse-services_8c-example.html
+ * which is also LGPL 2.1 or later.
+ *
+ * */
+
+#include "iio-lock.h"
+#include "iio-private.h"
+#include "network.h"
+
+#include "debug.h"
+
+/* Some functions for handling common linked list operations */
+static void dnssd_remove_node(struct dns_sd_discovery_data **ddata, int n)
+{
+
+	struct dns_sd_discovery_data *d, *ndata, *ldata, *tdata;
+	int i;
+
+	d = *ddata;
+
+	if (n == 0) {
+		tdata = d->next;
+		dnssd_free_discovery_data(d);
+		d = tdata;
+	} else {
+		for (i = 0, ndata = d; ndata->next != NULL; ndata = ndata->next) {
+			if (i == n) {
+				/* Could be NULL or node, both are OK */
+				tdata = ndata->next;
+				/* free the node to be removed */
+				dnssd_free_discovery_data(ndata);
+				/* snip it out */
+				ldata->next = tdata;
+				break;
+			}
+			ldata = ndata;
+			i++;
+		}
+	}
+
+	*ddata = d;
+}
+
+/*
+ * remove the ones in the list that you can't connect to
+ * This is sort of silly, but we have seen non-iio devices advertised
+ * and discovered on the network. Oh well....
+ */
+void port_knock_discovery_data(struct dns_sd_discovery_data **ddata)
+{
+	struct dns_sd_discovery_data *d, *ndata;
+	int i, ret;
+
+	d = *ddata;
+	iio_mutex_lock(d->lock);
+	for (i = 0, ndata = d; ndata->next != NULL; ndata = ndata->next) {
+		char port_str[6];
+		struct addrinfo hints, *res, *rp;
+		int fd;
+
+		memset(&hints, 0, sizeof(hints));
+		hints.ai_family = AF_UNSPEC;
+		hints.ai_socktype = SOCK_STREAM;
+
+		iio_snprintf(port_str, sizeof(port_str), "%hu", ndata->port);
+		ret = getaddrinfo(ndata->addr_str, port_str, &hints, &res);
+
+		/* getaddrinfo() returns a list of address structures */
+		if (ret) {
+			DEBUG("Unable to find host ('%s'): %s\n",
+					ndata->hostname,
+					gai_strerror(ret));
+			dnssd_remove_node(&d, i);
+		} else {
+			for (rp = res; rp != NULL; rp = rp->ai_next) {
+				fd = create_socket(rp, DEFAULT_TIMEOUT_MS);
+				if (fd < 0) {
+					DEBUG("Unable to open %s%s socket ('%s:%d' %s)\n",
+							res->ai_family == AF_INET ? "ipv4" : "",
```

Comment:
should this be  `rp` vs `res` ?
i.e. `res->ai_family == AF_INET` ?

not a big one, since it's mostly debugging print;

Reply:
good eye. Thanks - fixed.